Update containerd client and dependencies to v1.2.0-rc.1

thaJeztah opened the pull request:
Full diffs and included changes:

- https://github.com/containerd/containerd/compare/d97a907f7f781c0ab8340877d8e6b53cc7f1c2f6...99fc40fd6088baebe3c18b8f8093cd50d58815d6
  - https://github.com/containerd/containerd/pull/2633 Support importing docker images
  - https://github.com/containerd/containerd/pull/2663 Reduce publish connection timeout
  - https://github.com/containerd/containerd/pull/2664 Add nosuid,noexec,nodev to proc
  - https://github.com/containerd/containerd/pull/2640 Support parallel decompression (pigz)
  - https://github.com/containerd/containerd/pull/2649 Add flag to ctr for running with "NoNewPrivileges: false"

full diff since rc.0: https://github.com/containerd/containerd/compare/v1.2.0-rc.0...v1.2.0-rc.1

- https://github.com/opencontainers/runtime-spec/compare/d810dbc60d8c5aeeb3d054bd1132fab2121968ce...eba862dc2470385a233c7507392675cbeadf7353
  - https://github.com/opencontainers/runtime-spec/pull/989 Support for network namespace in windows 
- https://github.com/opencontainers/selinux/compare/b29023b86e4a69d1b46b7e7b4e2b6fda03f0b9cd...b6fa367ed7f534f9ba25391cc2d467085dbb445a
  - https://github.com/opencontainers/selinux/pull/25 Support SELinux systems without MCS Labels
  - https://github.com/opencontainers/selinux/pull/26 getSELinuxfs: fix/improve/speedup plus test fixes
  - https://github.com/opencontainers/selinux/pull/28 selinux: minor and idiomatic cleanups
  - https://github.com/opencontainers/selinux/pull/29 SELinux needs to be able to reset process label to "" 
  - https://github.com/opencontainers/selinux/pull/31 go-selinux is failing to build on non linux systems
  - https://github.com/opencontainers/selinux/pull/33 go-selinux/selinux.go: fix return value for FSCreateLabel
  - https://github.com/opencontainers/selinux/pull/30 Add additional directories to block global relabel
  - https://github.com/opencontainers/selinux/pull/34 Another missing interface for non linux
  - https://github.com/opencontainers/selinux/pull/32 go-selinux: Use selinux build tag in go-selinux/
  - https://github.com/opencontainers/selinux/pull/19 Add CanonicalizeContext function

- github.com/containerd/continuity https://github.com/containerd/continuity/compare/f44b615e492bdfb371aae2f76ec694d9da1db537...bd77b46c8352f74eb12c85bdc01f4b90f69d66b4
  - https://github.com/containerd/continuity/pull/123 driver/{Mknod,Mkfifo,Lchmod}: return PathError
  - https://github.com/containerd/continuity/pull/133 Switch from x/net/context to context
  - https://github.com/containerd/continuity/pull/129 Add fileheaders with ltag tool
  - https://github.com/containerd/continuity/pull/134 Remove unreachable block in fs path cleanup
- github.com/opencontainers/runc https://github.com/opencontainers/runc/compare/20aff4f0488c6d4b8df4d85b4f63f1f704c11abd...00dc70017d222b178a002ed30e9321b12647af2d
  - https://github.com/opencontainers/runc/pull/1817 Stop relying on number of systems for cgroups
- github.com/Microsoft/hcsshim https://github.com/Microsoft/hcsshim/compare/v0.7.3...v0.7.6
  - https://github.com/Microsoft/hcsshim/pull/311 Adding Namespace to HNS exports
  - https://github.com/Microsoft/hcsshim/pull/309 Fix logging in CreateLayer
- github.com/containerd/ttrpc https://github.com/containerd/ttrpc/compare/94dde388801693c54f88a6596f713b51a8b30b2d...2a805f71863501300ae1976d29f0454ae003e85a
  - https://github.com/containerd/ttrpc/pull/29 Add client side context.Done support

kolyshkin commented:
The runc bump apparently breaks docker-in-lxd (proposed fix: https://github.com/opencontainers/runc/pull/1862) // cc @AkihiroSuda

thaJeztah commented:
Looks like this failure is pretty consistent on all Linux runs;

```
13:40:12 --- FAIL: TestContainerStartOnDaemonRestart (3.01s)
13:40:12 	daemon.go:290: [da6bebf524c56] waiting for daemon to start
13:40:12 	daemon.go:322: [da6bebf524c56] daemon started
13:40:12 	daemon.go:290: [da6bebf524c56] waiting for daemon to start
13:40:12 	daemon.go:322: [da6bebf524c56] daemon started
13:40:12 	daemon_linux_test.go:65: assertion failed: error is not nil: Error response from daemon: monitor task: cgroup is already being collected: unknown: failed to start test container
13:40:12 	daemon.go:280: [da6bebf524c56] exiting daemon
```

thaJeztah committed:
Update containerd dependencies

This updates the containerd dependencies to match
the versions used by the vendored containerd version

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Update containerd client and dependencies to v1.2.0-rc.1

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@kolyshkin @AkihiroSuda - rebased this PR, and changed it to only update the vendored (client) code. I'll open a separate PR to update the containerd runtime/daemon and runc version

thaJeztah commented:
Ah, hm... build failing (on Windows); thought I tried it, but perhaps before some of these updates;

```
16:19:39 ..\..\vendor\github.com\containerd\containerd\images\archive\reference.go:22:2: cannot find package "github.com/containerd/cri/pkg/util" in any of:
16:19:39 	c:\go\src\github.com\docker\docker\vendor\github.com\containerd\cri\pkg\util (vendor tree)
16:19:39 	c:\go\src\vendor\github.com\containerd\cri\pkg\util
16:19:39 	C:\go\src\github.com\docker\docker\vendor\github.com\containerd\cri\pkg\util
16:19:39 	C:\go\src\vendor\github.com\containerd\cri\pkg\util
16:19:39 	c:\go\src\github.com\containerd\cri\pkg\util (from $GOROOT)
16:19:39 	C:\go\src\github.com\containerd\cri\pkg\util (from $GOPATH)
```

thaJeztah commented on vendor/github.com/containerd/containerd/images/archive/reference.go:
```diff
@@ -0,0 +1,86 @@
+
+func normalizeReference(ref string) (string, error) {
+	// TODO: Replace this function to not depend on reference package
+	normalized, err := util.NormalizeImageRef(ref)
```
Due to this line, containerd/cri is needed as a dependency. Looking at that code, it's a utility that's only a thin wrapper around code from docker/distribution. Might be good to refactor that and either move it to docker/distribution or copy the utility function;

```go
// NormalizeImageRef normalizes the image reference following the docker convention. This is added
 // mainly for backward compatibility.
 // The reference returned can only be either tagged or digested. For reference contains both tag
 // and digest, the function returns digested reference, e.g. docker.io/library/busybox:latest@
 // sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa will be returned as
 // docker.io/library/busybox@sha256:7cc4b5aefd1d0cadf8d97d4350462ba51c694ebca145b08d7d41b41acc8db5aa.
 func NormalizeImageRef(ref string) (reference.Named, error) {
 	named, err := reference.ParseNormalizedNamed(ref)
 	if err != nil {
 		return nil, err
 	}
 	if _, ok := named.(reference.NamedTagged); ok {
 		if canonical, ok := named.(reference.Canonical); ok {
 			// The reference is both tagged and digested, only
 			// return digested.
 			newNamed, err := reference.WithName(canonical.Name())
 			if err != nil {
 				return nil, err
 			}
 			newCanonical, err := reference.WithDigest(newNamed, canonical.Digest())
 			if err != nil {
 				return nil, err
 			}
 			return newCanonical, nil
 		}
 	}
 	return reference.TagNameOnly(named), nil
 }
```

thaJeztah commented:
Updating containerd runtime in https://github.com/moby/moby/pull/37932

kolyshkin commented:
LGTM, thanks @thaJeztah

cpuguy83 commented:
Are we desiring to merge the RC into master?

AkihiroSuda commented:
either way, runc is still RC 😎

thaJeztah commented:
@cpuguy83 @AkihiroSuda @vdemeester It's green.. good to go?

vdemeester reviewed (approved):
LGTM :frog:

tiborvass commented:
@thaJeztah Actually, I don't think we want to vendor containerd rc2 in the code. We want master. RC2 is for the binary.

Srinidhi2301 commented:
hello everyone,
I do checkpoint restore on docker ( Tried on 18. ,19. , 20. versions)/ containerd (tried using several versions)
criu version: 3.11
ubuntu: 18.03
bionic arm64

Tried various combinations in version for docker and containerd from the below link

https://download.docker.com/linux/ubuntu/dists/bionic/pool/stable/arm64/

But still not successful. 

In some versions, I get the below error,
Error response from daemon: failed to retrieve OCI runtime container pid: open /run/containerd/io.containerd.runtime.v1.linux/moby/1a09d8379341aba1f401a0944b52722b32edfbc56ac7b68010bcd5e5e0ae38e4/init.pid: no such file or directory: unknown

In some versions, the containers which i try to start from checkpoint does not start and print no logs.

As suggested above, I tried to use containerd containerd.io_1.2.0~rc.2-1_arm64.deb
but i was getting the below error,
Error response from daemon: failed to retrieve OCI runtime container pid: open /run/containerd/io.containerd.runtime.v1.linux/moby/1a09d8379341aba1f401a0944b52722b32edfbc56ac7b68010bcd5e5e0ae38e4/init.pid: no such file or directory: unknown

could anyone please help! Thanks in advance
